File: src/2022/10.ts
// https://adventofcode.com/2022/day/10

export class Solver {
  private _input: string[];
  private _commands: number[];

  constructor(input: string[]) {
    this._input = input;
  }

  private _buildCommands(): void {
    this._commands = [];

    this._input.forEach(line => {
      const addMatch = line.match(/^addx (-?\d+)/);

      if (addMatch) {
        const value = parseInt(addMatch[1]);
        this._commands.push(0);
        this._commands.push(value);
      } else {
        this._commands.push(0);
      }
    });
  }

  private _renderPixelMatrix(pixelMatrix: boolean[][]): string {
    const crt_width = 40;
    const crt_height = 6;

    let lines = [];

    for (let i = 0; i < crt_height; i++) {
      lines[i] = "";

      for (let j = 0; j < crt_width; j++) {
        if (pixelMatrix[i]?.[j] === true) {
          lines[i] += "#";
        } else {
          lines[i] += ".";
        }
      }
    }

    return lines.join("\n");
  }

  public part1(): number {
    this._buildCommands();

    let cycle = 0;
    let x = 1;

    let signalStrengthSum = 0;

    this._commands.forEach(command => {
      cycle++;

      if ((cycle - 20) % 40 === 0) {
        signalStrengthSum += cycle * x;
      }

      x += command;
    });

    return signalStrengthSum;
  }

  public part2(): string {
    this._buildCommands();

    let cycle = 0;
    let x = 1;

    let pixelMatrix: boolean[][] = [];
    let i = 0;
    let j = 0;

    this._commands.forEach(command => {
      cycle++;

      if (Math.abs(x - j) <= 1) {
        pixelMatrix[i] ||= [];
        pixelMatrix[i][j] = true;
      }

      if (cycle % 40 === 0) {
        // Start a new row
        j = 0;
        i++;
      } else {
        // Move forward within the current row
        j++;
      }

      x += command;
    });

    return this._renderPixelMatrix(pixelMatrix);
  }
}

File: rust/2022/src/bin/10.rs
// https://adventofcode.com/2022/day/10

use regex::Regex;

const CRT_WIDTH: usize = 40;
const CRT_HEIGHT: usize = 6;

fn build_commands(input: &str) -> Vec<i16> {
    let re = Regex::new(r"^addx (-?\d+)").unwrap();
    let mut commands = vec![];

    input.lines().for_each(|line| {
        let caps = re.captures(line);

        if let Some(add_command) = caps {
            let value = add_command[1].parse::<i16>().unwrap();
            commands.push(0);
            commands.push(value);
        } else {
            commands.push(0);
        }
    });

    commands
}

fn render_pixel_matrix(pixel_matrix: [[bool; CRT_WIDTH]; CRT_HEIGHT]) -> String {
    let mut output = String::with_capacity(CRT_WIDTH * CRT_HEIGHT + CRT_HEIGHT);

    for i in 0..CRT_HEIGHT {
        for j in 0..CRT_WIDTH {
            if pixel_matrix[i][j] {
                output.push('#');
            } else {
                output.push('.');
            }
        }

        if i < CRT_HEIGHT - 1 {
            output.push('\n');
        }
    }

    output
}

fn part1(input: &str) -> i16 {
    let mut cycle = 0;
    let mut x = 1;
    let mut signal_strength_sum = 0;

    build_commands(input).iter().for_each(|command| {
        cycle += 1;

        if (cycle - 20) % 40 == 0 {
            signal_strength_sum += cycle * x;
        }

        x += command;
    });

    signal_strength_sum
}

fn part2(input: &str) -> String {
    let mut cycle = 0;
    let mut x = 1;
    let mut pixel_matrix = [[false; CRT_WIDTH]; CRT_HEIGHT];

    let mut i = 0;
    let mut j = 0;

    build_commands(input).iter().for_each(|command| {
        cycle += 1;

        if (x - j as i16).abs() <= 1 {
            pixel_matrix[i][j] = true;
        }

        if cycle % 40 == 0 {
            // Start a new row
            j = 0;
            i += 1;
        } else {
            // Move forward within the current row
            j += 1;
        }

        x += command;
    });

    render_pixel_matrix(pixel_matrix)
}

fn main() {
    let input = include_str!("../../../../_input/2022/10.txt").trim_end();

    println!("part 1: {}", part1(input));
    println!("part 2:\n{}", part2(input));
}

#[cfg(test)]
mod tests {
    const INPUT: &str = include_str!("../../../../_input/2022/10.txt");

    #[test]
    fn part1() {
        assert_eq!(super::part1(INPUT.trim_end()), 11820);
    }

    #[test]
    fn part2() {
        assert_eq!(
            super::part2(INPUT.trim_end()),
            "####.###....##.###..###..#..#..##..#..#.
#....#..#....#.#..#.#..#.#.#..#..#.#..#.
###..#..#....#.###..#..#.##...#..#.####.
#....###.....#.#..#.###..#.#..####.#..#.
#....#....#..#.#..#.#.#..#.#..#..#.#..#.
####.#.....##..###..#..#.#..#.#..#.#..#."
        );
    }
}
